macrosynergy/pnl/transaction_costs.py

```diff
@@ -0,0 +1,255 @@
+        return TransactionCosts(df=df, fids=get_fids(df), **self.DEFAULT_ARGS)
+
+    def __init__(
+        self,
+        df: QuantamentalDataFrame,
+        fids: List[str],
+        tcost_n: str,
+        rcost_n: str,
+        size_n: str,
+        tcost_l: str,
+        rcost_l: str,
+        size_l: str,
+    ) -> None:
+        self.sparse_costs = SparseCosts(df)
+        check_df_for_txn_stats(
+            df, fids, tcost_n, rcost_n, size_n, tcost_l, rcost_l, size_l
+        )
+        self.fids = sorted(set(fids))
+        self.tcost_n = tcost_n
+        self.rcost_n = rcost_n
+        self.size_n = size_n
+        self.tcost_l = tcost_l
+        self.rcost_l = rcost_l
+        self.size_l = size_l
+        self._txn_stats = [tcost_n, rcost_n, size_n, tcost_l, rcost_l, size_l]
+
+    def get_costs(self, fid: str, real_date: str) -> pd.Series:
+        assert fid in self.fids
+        assert hasattr(self, "sparse_costs") and hasattr(self.sparse_costs, "df_wide")
+        return self.sparse_costs.get_costs(fid=fid, real_date=real_date)
+
+    @staticmethod
+    def extrapolate_cost(
+        trade_size: Number,
+        median_size: Number,
+        median_cost: Number,
+        pct90_size: Number,
+        pct90_cost: Number,
+    ) -> Number:
+        return extrapolate_cost(
+            trade_size=trade_size,
+            median_size=median_size,
+            median_cost=median_cost,
+            pct90_size=pct90_size,
+            pct90_cost=pct90_cost,
+        )
+
+    def bidoffer(self, fid: str, trade_size: Number, real_date: str) -> Number:
+        row = self.sparse_costs.get_costs(fid=fid, real_date=real_date)
+        d = dict(
+            trade_size=trade_size,
+            median_size=row[fid + self.size_n],
+            median_cost=row[fid + self.tcost_n],
+            pct90_size=row[fid + self.size_l],
+            pct90_cost=row[fid + self.tcost_l],
+        )
+        return self.extrapolate_cost(**d)
+
+    def rollcost(self, fid: str, trade_size: Number, real_date: str) -> Number:
+        row = self.sparse_costs.get_costs(fid=fid, real_date=real_date)
+        d = dict(
+            trade_size=trade_size,
+            median_size=row[fid + self.size_n],
+            median_cost=row[fid + self.rcost_n],
+            pct90_size=row[fid + self.size_l],
+            pct90_cost=row[fid + self.rcost_l],
+        )
+        return self.extrapolate_cost(**d)
+
+
+class ExampleAdapter(TransactionCosts):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+    @staticmethod
+    def extrapolate_cost(
+        trade_size: Number,
+        median_size: Number,
+        median_cost: Number,
+        pct90_size: Number,
+        pct90_cost: Number,
+    ) -> Number:
+        # just as an example
+        u = trade_size * median_cost * median_cost / (median_size * median_size)
+        v = trade_size * pct90_cost * pct90_cost / (pct90_size * pct90_size)
+        return (u + v) / 2
```

Comment:
Equivalent to:
```python
u = median_cost * median_cost / (median_size * median_size)
v = pct90_cost * pct90_cost / (pct90_size * pct90_size)
avg_costs = (u+ v) /2
return trade_size * avg_costs
```
unless I am mistaken.

Reply:
Why cost and sized squared? Wouldn't this be better:

```python
u = median_cost / median_size
v = pct90_cost / pct90_size 
avg_costs = (u+ v) /2
return trade_size * avg_costs
```

Reply:
Mostly because it was just an example.
https://github.com/macrosynergy/macrosynergy/blob/adfea08c26305b4ecbe03ea326d51710c380ce9a/macrosynergy/pnl/transaction_costs.py#L206 

Have updated it with your example:
> 
> ```python
> u = median_cost / median_size
> v = pct90_cost / pct90_size 
> avg_costs = (u+ v) /2
> return trade_size * avg_costs
> ```